postmark.go

```diff
@@ -0,0 +1,185 @@
+type Postmark interface {
+	SetClient(client *http.Client) Postmark
+
+	// Email sends a single email with custom content.
+	// http://developer.postmarkapp.com/developer-api-email.html#send-email
+	Email(ctx context.Context, email *Email) error
+
+	// EmailWithTemplate sends an email with templated content.
+	// http://developer.postmarkapp.com/developer-api-templates.html#email-with-template
+	EmailWithTemplate(ctx context.Context, email *EmailWithTemplate) error
+}
+
+type postmark struct {
+	serverToken  string
+	accountToken string
+	client       *http.Client
+}
+
+// New returns an initialized Postmark client
+func New(serverToken, accountToken string) Postmark {
+	return &postmark{
+		serverToken:  serverToken,
+		accountToken: accountToken,
+	}
+}
+
+func (p *postmark) exec(ctx context.Context, r *http.Request) error {
+	r.Header.Set("Accept", "application/json")
+	r.Header.Set("Content-Type", "application/json")
+	r.Header.Set("X-Postmark-Server-Token", p.serverToken)
+
+	resp, err := p.httpclient().Do(r)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	var er EmailResponse
+	er.StatusCode = resp.StatusCode
+
+	if err := json.NewDecoder(resp.Body).Decode(&er); err != nil {
+		return err
+	}
+
+	if resp.StatusCode/100 != 2 {
+		if er.IsError() {
+			return &er
+		}
+		return fmt.Errorf("postmark call errored with status: %d", resp.StatusCode)
+	}
+
+	if er.IsError() {
+		return &er
+	}
+	return nil
+}
+
+func (p *postmark) httpclient() *http.Client {
+	if p.client != nil {
+		return p.client
+	}
+	return http.DefaultClient
+}
+
+func (p *postmark) SetClient(client *http.Client) Postmark {
+	p.client = client
+	return p
+}
+
+// BaseEmail defines the fields common to all Postmark emails
+type BaseEmail struct {
+	From        string       `json:",omitempty"`
+	To          string       `json:",omitempty"`
+	Cc          string       `json:",omitempty"`
+	Bcc         string       `json:",omitempty"`
+	Tag         string       `json:",omitempty"`
+	ReplyTo     string       `json:",omitempty"`
+	Headers     []Header     `json:",omitempty"`
+	TrackOpens  bool         `json:",omitempty"`
+	Attachments []Attachment `json:",omitempty"`
+}
+
+// Header defines an email header within the Postmark API
+type Header struct {
+	Name  string
+	Value string
+}
+
+// Attachment defines an email attachment within the Postmark API
+type Attachment struct {
+	Name        string
+	Content     string
+	ContentType string
+}
+
+// EmailResponse is the response from the postmark API after an email is sent.
+// This can also be an error type for unsuccesful calls.
+type EmailResponse struct {
+	To          string
+	SubmittedAt time.Time
+	MessageID   string
+	ErrorCode   int
+	Message     string
+
+	// the HTTP status code of the response itself
+	StatusCode int `json:"-"`
+}
+
+// IsError returns whether or not the response indicated an error
+func (e *EmailResponse) IsError() bool {
+	return e.ErrorCode != 0
```

Comment:
Is there a situation in which an error would cause `e.ErrorCode` to be left at its initialized state? What do you think about a check for success in addition to checking for an error code?